core/engine/src/value/conversions/coerce.rs

```diff
@@ -0,0 +1,128 @@
+//! Types and functions for applying Coercing rules to [`JsValue`] when
+//! converting.
+
+use boa_engine::JsNativeError;
+
+use crate::value::TryFromJs;
+use crate::{Context, JsResult, JsString, JsValue};
+
+/// A wrapper type that allows coercing a `JsValue` to a specific type.
+/// This is useful when you want to coerce a `JsValue` to a Rust type.
+///
+/// # Example
+/// Convert a string to number.
+/// ```
+/// # use boa_engine::{Context, js_string, JsValue};
+/// # use boa_engine::value::{Convert, TryFromJs};
+/// # let mut context = Context::default();
+/// let value = JsValue::from(js_string!("42"));
+/// let Convert(coerced): Convert<i32> = Convert::try_from_js(&value, &mut context).unwrap();
+///
+/// assert_eq!(coerced, 42);
+/// ```
+///
+/// Convert a number to a bool.
+/// ```
+/// # use boa_engine::{Context, js_string, JsValue};
+/// # use boa_engine::value::{Convert, TryFromJs};
+/// # let mut context = Context::default();
+/// let value0 = JsValue::Integer(0);
+/// let value1 = JsValue::Integer(1);
+/// let value_nan = JsValue::Rational(f64::NAN);
+/// let Convert(coerced0): Convert<bool> = Convert::try_from_js(&value0, &mut context).unwrap();
+/// let Convert(coerced1): Convert<bool> = Convert::try_from_js(&value1, &mut context).unwrap();
+/// let Convert(coerced_nan): Convert<bool> = Convert::try_from_js(&value_nan, &mut context).unwrap();
+///
+/// assert_eq!(coerced0, false);
+/// assert_eq!(coerced1, true);
+/// assert_eq!(coerced_nan, false);
+/// ```
```

Comment:
Nit: It would be good to remove all references to "coerce" or "coercion" in both the docs and the variable names, for consistency. Not marking this as a critical change because it's not really important.

Reply:
Done, I've also added a link to the section 9 of the spec that explains the conversion rules.